### PR TITLE
Fix #10871, null deref issue in afta

### LINKED ISSUE #10871: "aaa" crashes against (big endian?) MIPS binaries
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Gentoo x86 64
| File format of the file you reverse (mandatory)      | ELF
| Architecture/bits of the file (mandatory)            | MIPS 64
| r2 -v full output, **not truncated** (mandatory)     | radare2 2.8.0-git 18893 @ linux-x86-64 git.2.7.0-200-gdcdb33bda
commit: dcdb33bda344e37e2cd6981b902dd2e71b71b806 build: 2018-07-31__11:49:01

### Expected behavior

Unknown, but not a segfault.

### Actual behavior

```
$ R_DEBUG=yes radare2 -A sgitest.elf 
Cannot open /home/nyef/bin/prefix/radare2/lib/radare2/last
Cannot open /home/nyef/bin/prefix/radare2/lib/radare2/2.8.0-git
Cannot open directory '/home/nyef/.local/share/radare2/plugins'
Cannot open directory '/home/nyef/bin/prefix/radare2/lib/radare2-extras/2.8.0-git'
Cannot open directory '/home/nyef/bin/prefix/radare2/lib/radare2-bindings/2.8.0-git'
Warning: Cannot initialize program headers
Warning: Cannot initialize dynamic strings
Warning: Cannot initialize dynamic section
Warning: run r2 with -e io.cache=true to fix relocations in disassembly
[x] Analyze all flags starting with sym. and entry0 (aa)
[ ] 
[Value from 0x08000000 to 0x0800e798
aav: 0x08000000-0x0800e798 in 0x8000000-0x800e798
aav: 0x08000000-0x0800e798 in 0x800e7a0-0x8010731
aav: 0x08000000-0x0800e798 in 0x8010bb8-0x80265a0
Value from 0x0800e7a0 to 0x08010731
aav: 0x0800e7a0-0x08010731 in 0x8000000-0x800e798
aav: 0x0800e7a0-0x08010731 in 0x800e7a0-0x8010731
aav: 0x0800e7a0-0x08010731 in 0x8010bb8-0x80265a0
Value from 0x08010bb8 to 0x080265a0
aav: 0x08010bb8-0x080265a0 in 0x8000000-0x800e798
aav: 0x08010bb8-0x080265a0 in 0x800e7a0-0x8010731
aav: 0x08010bb8-0x080265a0 in 0x8010bb8-0x80265a0
[x] Analyze function calls (aac)
[x] Analyze len bytes of instructions for references (aar)
[x] Constructing a function name for fcn.* and sym.func.* functions (aan)
[Backtrace 10 stack frames.for all functions (afta)
/home/nyef/bin/prefix/radare2/lib/libr_util.so(r_sys_backtrace+0x28)[0x7f8d564f78f0]
/home/nyef/bin/prefix/radare2/lib/libr_util.so(+0x4da72)[0x7f8d564f7a72]
/lib64/libpthread.so.0(+0x144e0)[0x7f8d5629e4e0]
/lib64/libc.so.6(+0x9d996)[0x7f8d55f65996]
/lib64/libc.so.6(__strdup+0x1e)[0x7f8d55f54e1e]
/home/nyef/bin/prefix/radare2/lib/libr_core.so(+0x1817d1)[0x7f8d5a28e7d1]
/home/nyef/bin/prefix/radare2/lib/libr_core.so(r_core_anal_type_match+0xafc)[0x7f8d5a28ff97]
/home/nyef/bin/prefix/radare2/lib/libr_core.so(+0x9625e)[0x7f8d5a1a325e]
/home/nyef/bin/prefix/radare2/lib/libr_core.so(+0x9c9c6)[0x7f8d5a1a99c6]
/home/nyef/bin/prefix/radare2/lib/libr_core.so(+0xae72a)[0x7f8d5a1bb72a]
GNU gdb (Gentoo 8.1 p1) 8.1
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://bugs.gentoo.org/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
Attaching to process 29017
Reading symbols from /home/nyef/src/radare2/binr/radare2/radare2...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_core.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_parse.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_search.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_cons.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_config.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_bin.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_debug.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_anal.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_reg.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_bp.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_io.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_fs.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_asm.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_syscall.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_hash.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_magic.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_flag.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_egg.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_crypto.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_util.so...done.
Reading symbols from /lib64/libpthread.so.0...(no debugging symbols found)...done.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Reading symbols from /lib64/libc.so.6...(no debugging symbols found)...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_lang.so...done.
Reading symbols from /home/nyef/bin/prefix/radare2/lib/libr_socket.so...done.
Reading symbols from /lib64/libm.so.6...(no debugging symbols found)...done.
Reading symbols from /lib64/libdl.so.2...(no debugging symbols found)...done.
Reading symbols from /lib64/ld-linux-x86-64.so.2...(no debugging symbols found)...done.
Reading symbols from /lib64/libutil.so.1...(no debugging symbols found)...done.
Reading symbols from /usr/lib/gcc/x86_64-pc-linux-gnu/6.4.0/libgcc_s.so.1...(no debugging symbols found)...done.
0x00007f8d55f93dee in waitpid () from /lib64/libc.so.6
(gdb) back
#0  0x00007f8d55f93dee in waitpid () from /lib64/libc.so.6
#1  0x00007f8d55f0be43 in ?? () from /lib64/libc.so.6
#2  0x00007f8d5650d000 in r_sandbox_system (
    x=0x7ffedfa39560 "gdb --pid 29017", n=1) at sandbox.c:152
#3  0x00007f8d564f86af in r_sys_cmd (str=0x7ffedfa39560 "gdb --pid 29017")
    at sys.c:679
#4  0x00007f8d564f7a81 in signal_handler (signum=11) at sys.c:335
#5  <signal handler called>
#6  0x00007f8d55f65996 in ?? () from /lib64/libc.so.6
#7  0x00007f8d55f54e1e in strdup () from /lib64/libc.so.6
#8  0x00007f8d5a28e7d1 in get_src_regname (core=0x555dc7cd5580 <r>, 
    addr=134219732, regname=0x7ffedfa3a420 "", size=10) at anal_tp.c:135
#9  0x00007f8d5a28ff97 in r_core_anal_type_match (core=0x555dc7cd5580 <r>, 
    fcn=0x555dc9f56290) at anal_tp.c:486
#10 0x00007f8d5a1a325e in type_cmd (core=0x555dc7cd5580 <r>, 
    input=0x555dc9f25233 "a") at cmd_anal.c:685
#11 0x00007f8d5a1a99c6 in cmd_anal_fcn (core=0x555dc7cd5580 <r>, 
    input=0x555dc9f25231 "fta") at cmd_anal.c:2330
#12 0x00007f8d5a1bb72a in cmd_anal (data=0x555dc7cd5580 <r>, 
    input=0x555dc9f25231 "fta") at cmd_anal.c:7102
#13 0x00007f8d5a249c88 in r_cmd_call (cmd=0x555dc9bd4df0, 
    input=0x555dc9f25230 "afta") at cmd_api.c:237
#14 0x00007f8d5a1ffeab in r_core_cmd_subst_i (core=0x555dc7cd5580 <r>, 
---Type <return> to continue, or q <return> to quit---
    cmd=0x555dc9f25230 "afta", colon=0x0, tmpseek=0x7ffedfa3acd6) at cmd.c:2910
#15 0x00007f8d5a1fca22 in r_core_cmd_subst (core=0x555dc7cd5580 <r>, 
    cmd=0x555dc9f25230 "afta") at cmd.c:1939
#16 0x00007f8d5a202612 in r_core_cmd (core=0x555dc7cd5580 <r>, 
    cstr=0x7f8d5a2af017 "afta", log=0) at cmd.c:3614
#17 0x00007f8d5a202d64 in r_core_cmd0 (user=0x555dc7cd5580 <r>, 
    cmd=0x7f8d5a2af017 "afta") at cmd.c:3778
#18 0x00007f8d5a1ba7cf in cmd_anal_all (core=0x555dc7cd5580 <r>, 
    input=0x555dc9deaee2 "a") at cmd_anal.c:6831
#19 0x00007f8d5a1bb8a4 in cmd_anal (data=0x555dc7cd5580 <r>, 
    input=0x555dc9deaee1 "aa") at cmd_anal.c:7150
#20 0x00007f8d5a249c88 in r_cmd_call (cmd=0x555dc9bd4df0, 
    input=0x555dc9deaee0 "aaa") at cmd_api.c:237
#21 0x00007f8d5a1ffeab in r_core_cmd_subst_i (core=0x555dc7cd5580 <r>, 
    cmd=0x555dc9deaee0 "aaa", colon=0x0, tmpseek=0x7ffedfa3b326) at cmd.c:2910
#22 0x00007f8d5a1fca22 in r_core_cmd_subst (core=0x555dc7cd5580 <r>, 
    cmd=0x555dc9deaee0 "aaa") at cmd.c:1939
#23 0x00007f8d5a202612 in r_core_cmd (core=0x555dc7cd5580 <r>, 
    cstr=0x555dc7ad416a "aaa", log=0) at cmd.c:3614
#24 0x00007f8d5a202d64 in r_core_cmd0 (user=0x555dc7cd5580 <r>, 
    cmd=0x555dc7ad416a "aaa") at cmd.c:3778
#25 0x0000555dc7ad2298 in main (argc=3, argv=0x7ffedfa3b808, 
    envp=0x7ffedfa3b828) at radare2.c:1299
(gdb)
```
### Steps to reproduce the behavior 

`radare2 -A sgitest.elf`

### Additional Logs, screenshots, source-code,  configuration dump, ...

[sgitest.zip](https://github.com/radare/radare2/files/2246262/sgitest.zip)

### PR DESCRIPTION
**Pull request by sivaramaaa**

### CODE REVIEW

File: libr/core/anal_tp.c [hunk 1]
```diff
@@ -132,9 +132,12 @@ static void get_src_regname(RCore *core, ut64 addr, char *regname, int size) {
 	RRegItem *ri = r_reg_get (anal->reg, op_esil, -1);
 	if (ri) {
 		if ((anal->bits == 64) && (ri->size == 32)) {
-			char *foo = strdup (r_reg_32_to_64 (anal->reg, op_esil));
-			free (op_esil);
-			op_esil = foo;
+			const char *reg = r_reg_32_to_64 (anal->reg, op_esil);
+			if (reg) {
+				char *foo = strdup (reg);
```
Comment (ret2libc) [inline libr/core/anal_tp.c]: Could it be that we just need to add the register `zero` to the profile?

**Comment by radare** [comment]
Yeah. But that shouldnt crash

**Comment by ret2libc** [comment]
It shouldn’t but I’d prefer a eprintf that clearly says what is the problem more than hiding it

**Comment by radare** [comment]
Me too

**Comment by radare** [comment]
Merging as long as the crash is fixed. Improving it to make it work properly for mips may be fixed in a separate pr.

Thanks!

**Comment by ret2libc** [comment]
@sivaramaaa if you could also do the proper fix in a separate PR it would be great! Btw, as it is right now I'm not sure you even need the `foo` variable

**Comment by sivaramaaa** [comment]
> Could it be that we just need to add the register zero to the profile?

@ret2libc can u pls explain a littile bit more (sorry i am not very clear on this :{ )


> Btw, as it is right now I'm not sure you even need the foo variable

if i remove foo variable , and make changes like below , it would cause a leak no ?

```c
op_esil = strdup ( r_reg_32_to_64 (anal->reg, op_esil) )
```

**Comment by ret2libc** [comment]
so it seems to me, when running the faulting binary, that `r_reg_32_to_64` returns NULL because it tries to resolve the `zero` register. Now, I don't know all the details, but I think every architecture has a "profile" that defines all the registers and their details. I believe (not 100% sure) that the reason the function is returning NULL is just because on mips there are some wrong/missing info (in particular about this `zero` registry`.. I know nothing about MIPS)

So I would try to fix that first, before "ignoring" the issue.

Btw, what I meant is that you could do:
```
                        const char *reg = r_reg_32_to_64 (anal->reg, op_esil);
			if (reg) {
                                free (op_esil);
				op_esil = strdup (reg);
			}
```

**Comment by sivaramaaa** [comment]
@ret2libc thanks , now both issue is very clear to me , i will look into this !

**Pull request merged**